src/main.c

```diff
@@ -195,46 +175,32 @@ int main(void)
 		LOG_ERR("Unable to configure LED for Golioth Logo");
 	}
 
-	/* Initialize app state */
-	app_state_init(client);
-
-	/* Initialize app work */
-	app_work_init(client);
-
-	/* Initialize DFU components */
-	app_dfu_init(client);
-
-	/* Initialize app settings */
-	app_settings_init(client);
-
-	/* Initialize app RPC */
-	app_rpc_init(client);
-
-	/* Register Golioth on_connect callback */
-	client->on_connect = golioth_on_connect;
-
-	/* Start LTE asynchronously if the nRF9160 is used
+#ifdef CONFIG_SOC_NRF9160
+	/* Start LTE asynchronously if the nRF9160 is used.
 	 * Golioth Client will start automatically when LTE connects
 	 */
-	IF_ENABLED(CONFIG_SOC_NRF9160, (LOG_INF("Connecting to LTE, this may take some time...");
-					lte_lc_init_and_connect_async(lte_handler);));
 
+	LOG_INF("Connecting to LTE, this may take some time...");
+	register_lte_event_handler();
+	register_lte_on_connect_cb(process_lte_connected);
+
+#else
 	/* If nRF9160 is not used, start the Golioth Client and block until connected */
-	if (!IS_ENABLED(CONFIG_SOC_NRF9160)) {
-		/* Run WiFi/DHCP if necessary */
-		if (IS_ENABLED(CONFIG_GOLIOTH_SAMPLES_COMMON)) {
-			net_connect();
-		}
 
-		/* Start Golioth client */
-		golioth_system_client_start();
+	/* Run WiFi/DHCP if necessary */
+	if (IS_ENABLED(CONFIG_GOLIOTH_SAMPLES_COMMON)) {
+		net_connect();
+	}
 
-		/* Block until connected to Golioth */
-		k_sem_take(&connected, K_FOREVER);
+	/* Start Golioth client */
+	start_golioth_client();
 
-		/* Turn on Golioth logo LED once connected */
-		gpio_pin_set_dt(&golioth_led, 1);
-	}
+	/* Block until connected to Golioth */
+	k_sem_take(&connected, K_FOREVER);
```

Comment:
I think we want to make sure that the `on_client_event()` callback for the Reference Design ran before returning to program flow. If we blocked on `start_golioth_client()` it wouldn't guarantee. For the template it doesn't really matter, but for descendant reference designs it might.

Also, how would you block on that? I think a semaphore here is a better solution than repeatedly calling a function.

Reply:
That example is polling with a timeout. I do prefer the semaphore to that approach.